python/GafferSceneTest/VisibleSetTest.py

```diff
@@ -0,0 +1,148 @@
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of Cinesite VFX Ltd. nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import IECore
+
+import GafferSceneTest
+import GafferScene
+
+class VisibleSetTest( GafferSceneTest.SceneTestCase ) :
+
+	def testExpansions( self ) :
+	
+		e = GafferScene.VisibleSet()
+		e.expansions = IECore.PathMatcher( [ "/a", "/b", "/a/b", "/b/c", "/c/d/e" ] )
+
+		self.assertEqual( e.match( "/a" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/b" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/a/b" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/b/c" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/c/d/e" ), IECore.PathMatcher.Result.ExactMatch )
+
+		self.assertEqual( e.match( "/" ), IECore.PathMatcher.Result.NoMatch )
+		self.assertEqual( e.match( "/c" ), IECore.PathMatcher.Result.NoMatch )
+		self.assertEqual( e.match( "/c/d" ), IECore.PathMatcher.Result.NoMatch )
+		self.assertEqual( e.match( "/a/b/c" ), IECore.PathMatcher.Result.NoMatch )
+
+	def testInclusions( self ) :
+
+		e = GafferScene.VisibleSet()
+		e.inclusions = IECore.PathMatcher( [ "/a", "/b/c" ] )
+
+		# Paths specifically added as inclusions
+		self.assertEqual( e.match( "/a" ), IECore.PathMatcher.Result.ExactMatch )
+		self.assertEqual( e.match( "/b/c" ), IECore.PathMatcher.Result.ExactMatch )
+
+		# Paths with included descendants
+		self.assertEqual( e.match( "/" ), IECore.PathMatcher.Result.DescendantMatch )
+		self.assertEqual( e.match( "/b" ), IECore.PathMatcher.Result.DescendantMatch )
+
+		# Paths with included ancestors
+		self.assertEqual( e.match( "/a/b" ), IECore.PathMatcher.Result.AncestorMatch )
+		self.assertEqual( e.match( "/b/c/d/e" ), IECore.PathMatcher.Result.AncestorMatch )
```

Comment:
I think perhaps it's simpler for the caller if these include `ExactMatch`? The caller is thinking "should I render _this_?", and checking `ExactMatch` seems to be the way to answer that question.

Put another way, I think `match == AncestorMatch` should mean that this location _shouldn't_ be rendered, but at least one ancestor _should_ (for example, the parent isn't expanded, but the parent's-parent is).

Reply:
As above, acc72e6 has the refined match behaviour where `ExactMatch` is included for any location that is considered visible.

A non-visible location that happens to have an expanded ancestor doesn't currently return an `AncestorMatch`, but I think that should be simple enough to include if we think it necessary for completeness. I've also added a todo covering what we've discussed regarding adding an AllAncestorsMatch (or equivalent) to PathMatcher.